### ISSUE TITLE
False positive error for function instantiating rank-2 structural type 

### ISSUE THREAD

**Issue by robert-bn** [microsoft/pyright#8177]
I apologize if this is as designed or a duplicate of an existing issue.

The following code produces an error in pyright for `example_2`, but not for `example_1`. I would expect no error in either case, as I believe `make_getter` is a subtype of `PolymorphicListItemGetter`.
```python
from typing import Protocol


class Getter[T, U]: ...

class PolymorphicListItemGetter(Protocol):
    def __call__[T](self) -> Getter[list[T], T]: ...


def example_1() -> PolymorphicListItemGetter:
    class MakeGetter(PolymorphicListItemGetter):
        def __call__[T](self) -> Getter[list[T], T]: ...

    return MakeGetter()


def example_2() -> PolymorphicListItemGetter:
    def make_getter[T]() -> Getter[list[T], T]: ...

    return make_getter
```
The error reported by pyright is
```
error: Expression of type "() -> Getter[list[T@make_getter], T@make_getter]" is incompatible with return type "PolymorphicListItemGetter"
    Type "() -> Getter[list[T@make_getter], T@make_getter]" is incompatible with type "() -> Getter[list[T@__call__], T@__call__]"
      Function return type "Getter[list[T@make_getter], T@make_getter]" is incompatible with type "Getter[list[T@__call__], T@__call__]"
        "Getter[list[T@make_getter], T@make_getter]" is incompatible with "Getter[list[T@__call__], T@__call__]"
          Type parameter "T@Getter" is covariant, but "list[T@make_getter]" is not a subtype of "list[T@__call__]"
            "list[T@make_getter]" is incompatible with "list[T@__call__]"
          Type parameter "U@Getter" is covariant, but "T@make_getter" is not a subtype of "T@__call__"
            Type "T@make_getter" is incompatible with type "T@__call__" (reportReturnType)
```
tested with command-line pyright version 1.1.368

**Comment by erictraut** [comment]
Thanks for the bug report and the great repro steps. I agree this appears to be a bug. I'll investigate further. If you're looking for a workaround in the meantime, if you make `T` a class-scoped type parameter in `PolymorphicListItemGetter` rather than a method-scoped type parameter, it will type check without error.

```python
class PolymorphicListItemGetter[T](Protocol):
    def __call__(self) -> Getter[list[T], T]: ...
```

**Comment by erictraut** [comment]
This is included in pyright 1.1.369.
